### PR TITLE
Resolves building issues with Intel compilers (ifx/icx)

### PR DESCRIPTION
**Pull request by mgs-52**
Adding/Replacing missing prototypes. Remove unsupported/bogus compiler switches.

TYPE: Bug fix

KEYWORDS: Missing Prototypes

SOURCE:  Alexander Knyazev

DESCRIPTION OF CHANGES:
Problem:  Most C code written long ago.  No prototypes were used.

Solution:
Add missing prototypes.

LIST OF MODIFIED FILES: 

external/io_grib1/grib1_util/read_grib.c
external/io_grib1/grib1_util/write_grib.c
external/io_grib1/grib1_util/read_grib.h
external/io_grib1/MEL_grib1/FTP_getfile.c
external/io_grib1/MEL_grib1/gribfuncs.h
external/io_grib1/MEL_grib1/gribgetbds.c
external/io_grib1/MEL_grib1/upd_child_errmsg.c
external/io_grib1/MEL_grib1/gribhdr2file.c
external/io_grib1/MEL_grib1/init_enc_struct.c
external/io_grib1/MEL_grib1/init_gribhdr.c
external/io_grib1/MEL_grib1/init_struct.c
external/io_grib1/MEL_grib1/ld_enc_input.c
external/io_grib1/MEL_grib1/ld_enc_lookup.c
external/io_grib1/MEL_grib1/map_lvl.c
external/io_grib1/MEL_grib1/map_parm.c
external/io_grib1/MEL_grib1/prt_inp_struct.c
external/io_grib1/MEL_grib1/prt_badmsg.c
external/io_grib1/MEL_grib1/set_bytes.c
external/io_grib1/MEL_grib1/ld_grib_origctrs.c
external/io_grib1/gribmap.c
external/io_grib1/grib1_routines.c
external/io_grib1/gribmap.h
external/io_grib1/trim.c
external/io_grib1/trim.h
external/io_grib_share/get_region_center.c
external/io_grib_share/open_file.c
external/RSL_LITE/gen_comms.c
external/RSL_LITE/c_code.c
external/RSL_LITE/buf_for_proc.c
external/RSL_LITE/rsl_malloc.c
external/RSL_LITE/rsl_bcast.c
external/RSL_LITE/task_for_point.c
external/RSL_LITE/period.c
external/RSL_LITE/swap.c
external/RSL_LITE/cycle.c
external/RSL_LITE/rsl_lite.h
tools/protos.h
tools/reg_parse.c
tools/misc.c
tools/symtab_gen.c
tools/gen_config.c
tools/sym.c
tools/gen_streams.c
tools/gen_irr_diag.c
tools/sym.h
frame/pack_utils.c
arch/configure.defaults 

TESTS CONDUCTED:
1. Compilation with Intel oneAPI compiler works with these changes.
2. The Jenkins tests are all passing.

RELEASE NOTE: This PR fixed some c code by adding and replacing missing prototypes to allow Intel oneAPI compiler to work. Remove unsupported/bogus compiler switches.

### CODE REVIEW

**Comment by weiwangncar** [comment]
Jenkins test results:

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 23           24
    Number of Builds       : 60           57
    Number of Simulations  : 158           150        0
    Number of Comparisons  : 95           86        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None

**Comment by islas** [comment]
@mgs-52 This should target `develop`

File: tools/symtab_gen.c [hunk 1]
```diff
@@ -32,25 +32,25 @@ For a sample main or calling program see the end of this file.
...
 
 /*  commented out 2-29-90
 static char * symtab[HASHSIZE] ;	
 */
-
+/*
 void * malloc() ;
 void * calloc() ;
+*/
+/* char * symget(char *name, char *(*newnode)(), char  **(*nodename)(char *), sym_nodeptr *(*nodenext)(sym_nodeptr), char ** symtab, int flag);*/
```
Comment (islas) [inline tools/symtab_gen.c:51]: remove

**Review by islas — CHANGES_REQUESTED** [review]

File: tools/symtab_gen.c (hunk 1 above)
Comment (mgs-52) [reply tools/symtab_gen.c:51]: Done.

File: frame/pack_utils.c
[the remark this answers is not in the document]
Comment (mgs-52) [reply frame/pack_utils.c:1]: Done

[the remark this answers is not in the document]
Comment (islas) [reply frame/pack_utils.c:1]: @mgs-52 Did you forget to push your changes?

[the remark this answers is not in the document]
Comment (mgs-52) [reply frame/pack_utils.c:1]: Hmm. Apparently :)

**Review by islas — DISMISSED** [review]

**Comment by weiwangncar** [comment]
@islas @mgduda I just compiled this code on Derecho. With these modifications, one can compile WRF using the default modules. Would it be ok if we move this PR for release-4.5.2 release?

**Comment by islas** [comment]
Yes, this should go into the next release. My own testing with the latest Intel compilers from their download page also works with these updates

**Comment by weiwangncar** [comment]
@mgs-52 Can you let us know your affiliation, so that we can fill the SOURCE section of the PR? Thanks.

**Comment by islas** [comment]
@mgs-52 Any updates? I think affiliation/SOURCE is the last change needed

**Comment by mgs-52** [comment]
Well. Technically I'm working in Intel. However this change I did on my own so fill free to Include Intel or not :).

**Review by islas — APPROVED** [review]

**Comment by islas** [comment]
@mgduda When you get a chance, could you take a look at this? Ignoring whitespace changes most of the edits are rearrangement of function order and fixing bad function signatures (implicit returns and Fortran-esque argument typing)

**Comment by ahheo** [comment]
@mgs-52 May you please do the same job for the codes under WRF/chem (especially those of KPP) as well? I guess WRF-chem users may also get trouble with ifx/icx. Actually, I have done the code modification dealing with the issues of missing of prototypes and function order, and successfully built the WRF with WRF_CHEM=1 and WRF_KPP=1 using the newest Intel compilers. Well, as this is of the same issue that this pull is dealing with, I therefore suggest to complete the same job taking into account WRF-chem before the pull to be merged. What do you think @weiwangncar @islas?

**Comment by weiwangncar** [comment]
@ahheo Can you point out the files you modified for ifx/icx in chem/ directory? I think it is also ok to do another PR to address WRF-Chem.

**Comment by ahheo** [comment]
@weiwangncar I have done another PR #1942 as you suggested.

**Review by weiwangncar — APPROVED** [review]

**Comment by smileMchen** [comment]
@mgs-52 Alexander Knyazev,
Would you please tell where you are working?  We would like to have this information for release note. Thanks.

**Comment by mgs-52** [comment]
> @mgs-52 Alexander Knyazev, Would you please tell where you are working? We would like to have this information for release note. Thanks.

Intel, but I as stated above this doesn't need to be included since I did it on my own.

**Comment by weiwangncar** [comment]
@mgs-52 Will these icx compiler flags, '-Wno-implicit-function-declaration -Wno-incompatible-function-pointer-types', help with any of the old c code changed in this PR? See PR-[1972](https://github.com/wrf-model/WRF/pull/1972) for WRFDA compilation change.

**Comment by mgs-52** [comment]
> @mgs-52 Will these icx compiler flags, '-Wno-implicit-function-declaration -Wno-incompatible-function-pointer-types', help with any of the old c code changed in this PR? See PR-[1972](https://github.com/wrf-model/WRF/pull/1972) for WRFDA compilation change. Just curious.
They do, but I'd say it's a clutch: better make the code "compatible" with  the ("C") standard rather than instruct compiler to ignore "incompatibilities".

**Pull request merged**